### PR TITLE
Feature/shelby financials account grouping2

### PR DESCRIPTION
**Pull request by guymaness2**
##### What does the change add or fix?

Added capability of separating transaction fees out into separate debit/credit lines. Also added a block setting to control grouping of accounts in the resulting Excel file.

**New Settings:**
Batch to Journal and Batches to Journal Block
* GL Account Grouping - Determines if debit and/or credit lines should be grouped and summed by GL account in the export file  

**New Attributes:**
Financial Gateway Attributes:  
* Shelby Financials Export > Gateway Fee Processing - Sets how the gateway should process fees
* Shelby Financials Export > Default Fee Account - Sets the default account for transaction fees
  
Financial Account Attributes:
* Shelby Financials Export > Transaction Fee Account - Sets the account to use for transaction fees

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

* Added support for handling transaction fees to FinancialGateways and FinancialAccounts
* Added block settings to group accounts in resulting Excel file. 

---------

##### Who reported, requested, or paid for the change?

Simple Donation

---------

##### Does this update or add options to the block UI?

New Financial Gateway Attributes:  
image
  
New Financial Account Attribute:  
image  
  
New block setting for Batch To Journal and Batches To Journal blocks:  
image
  
---------

##### What files does it affect?

* ShelbyFinancials/BatchToJournal.ascx.cs
* ShelbyFinancials/BatchesToJournal.ascx.cs
  
---------

##### Is it a breaking change for other versions/clients?

Has dependency on https://github.com/KingdomFirst/RockAssemblies/pull/123

### CODE REVIEW

File: ShelbyFinancials/BatchToJournal.ascx.cs [hunk 1]
```diff
@@ -38,15 +38,66 @@ namespace RockWeb.Plugins.rocks_kfs.ShelbyFinancials
...
+        "Close Batch",
+        Description = "Flag indicating if the Financial Batch be closed in Rock when successfully posted to Intacct.",
+        IsRequired = false,
+        DefaultBooleanValue = true,
+        Order = 1,
+        Key = AttributeKey.CloseBatch )]
+
+    [EnumField(
+        "GL Account Grouping",
+        Description = "Determines if debit and/or credit lines should be grouped and summed by GL account in the export file.",
```
Comment (guymaness2) [inline ShelbyFinancials/BatchToJournal.ascx.cs:59]:
```suggestion
        Description = "Determines if debit and/or credit lines should be grouped and summed by GL account in the export file. NOTE: Unique Projects, Regions, Funds, etc. may result in multiple lines even if account is grouped.",
```

File: ShelbyFinancials/BatchesToJournal.ascx.cs [hunk 2]
```diff
@@ -45,16 +45,75 @@ namespace RockWeb.Plugins.rocks_kfs.ShelbyFinancials
...
+        "Months Back",
+        Description = "The number of months back that batches should be loaded. This is helpful to prevent database timeouts if there are years of historical batches.",
+        IsRequired = true,
+        DefaultIntegerValue = 2,
+        Order = 2,
+        Key = AttributeKey.MonthsBack )]
+
+    [EnumField(
+        "GL Account Grouping",
+        Description = "Determines if debit and/or credit lines should be grouped and summed by GL account in the export file.",
```
Comment (guymaness2) [inline ShelbyFinancials/BatchesToJournal.ascx.cs:74]:
```suggestion
        Description = "Determines if debit and/or credit lines should be grouped and summed by GL account in the export file. NOTE: Unique Projects, Regions, Funds, etc. may result in multiple lines even if account is grouped.",
```

File: ShelbyFinancials/ReadMe.md [hunk 3]
```diff
@@ -40,28 +40,32 @@ After install, the Shelby Financials Batch to Journal block was added to your Ba
...
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">1</span> | **Detail Page** Link to the Financial Batch Details page |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">2</span> | **Button Text** Customize the text for the export button |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">3</span> | **Months Back** Number of months back that batches should be loaded. This is helpful to prevent database timeouts if there are years of historical batches |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">4</span> | **GL Account Grouping** Determines if debit and/or credit lines should be grouped and summed by GL account in the export file |
```
Comment (guymaness2) [inline ShelbyFinancials/ReadMe.md:66]: Add detailed note describing what fields are grouped by for debit grouping, credit grouping, and financial account grouping

File: ShelbyFinancials/ReadMe.md [hunk 4]
```diff
@@ -40,28 +40,32 @@ After install, the Shelby Financials Batch to Journal block was added to your Ba
...
-| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">3</span> | **Close Batch** Control whether the batch gets marked as closed in Rock after export
-
-<div style="page-break-after: always;"></div>
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">1</span> | **Button Text** Customize the text on the export button |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">2</span> | **Close Batch** Control whether the batch gets marked as closed in Rock after export |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">3</span> | **GL Account Grouping** Determines if debit and/or credit lines should be grouped and summed by GL account in the export file |
```
Comment (guymaness2) [inline ShelbyFinancials/ReadMe.md:51]: Add detailed note describing what fields are grouped by for debit grouping, credit grouping, and financial account grouping

File: ShelbyFinancials/ReadMe.md [hunk 5]
```diff
@@ -40,28 +40,32 @@ After install, the Shelby Financials Batch to Journal block was added to your Ba
...
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">3</span> | **Months Back** Number of months back that batches should be loaded. This is helpful to prevent database timeouts if there are years of historical batches |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">4</span> | **GL Account Grouping** Determines if debit and/or credit lines should be grouped and summed by GL account in the export file |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">5</span> | **Journal Description Lava** Lava for the journal description column per line. Default: Batch.Id: Batch.Name |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">5</span> | **Enable Debug** Outputs the object graph to help create your Lava syntax |
```
Comment (guymaness2) [inline ShelbyFinancials/ReadMe.md:68]:
```suggestion
| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">6</span> | **Enable Debug** Outputs the object graph to help create your Lava syntax |
```

File: ShelbyFinancials/ReadMe.md [hunk 6]
```diff
@@ -86,7 +90,21 @@ The export will always create (at a minimum) two lines for a Journal - a debit a
...
+
+#### Financial Gateway Attributes
+
+If your financial gateway reports transaction fees to Rock in their transaction download, you may want to configure these Financial Gateway attributes to choose how those fees are handled in your export file.
+
+![](../.screenshots/ShelbyFinancials/GatewayAttributes.png)
+
+| | |
+| --- | ---- |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">1</span> | **Gateway Fee Processing** How should the Intacct Export plugin process transaction fees? DEFAULT: No special handling of transaction fees will be performed. NET DEBIT: Add credit entries for any transaction fees and use net amount (amount - transaction fees) for debit account entries. GROSS DEBIT: Debit account entries are left untouched (gross) and new debit and credit entries will be added for any transaction fees. NOTE: Both Net Debit and Gross Debit require a Fee Account attribute be set on either the financial gateway or financial account. |
```
Comment (nateh777) [inline ShelbyFinancials/ReadMe.md:106]:
```suggestion
| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">1</span> | **Gateway Fee Processing** How should the Shelby Financials plugin process transaction fees? DEFAULT: No special handling of transaction fees will be performed. NET DEBIT: Add credit entries for any transaction fees and use net amount (amount - transaction fees) for debit account entries. GROSS DEBIT: Debit account entries are left untouched (gross) and new debit and credit entries will be added for any transaction fees. NOTE: Both Net Debit and Gross Debit require a Fee Account attribute be set on either the financial gateway or financial account. |
```

Comment (nateh777) [reply ShelbyFinancials/ReadMe.md:106]: Can you also try to see if you can make a bulleted list out of the items? If it isn't possible in a table that is fine, but I just want to try.

**Comment by guymaness2** [comment]
@nateh777 should be ready for merge

File: ShelbyFinancials/ReadMe.md [hunk 7]
```diff
@@ -48,7 +48,7 @@ After install, the Shelby Financials Batch to Journal block was added to your Ba
...
-| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">3</span> | **GL Account Grouping** Determines if debit and/or credit lines should be grouped and summed by GL account in the export file |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">3</span> | **GL Account Grouping** Determines if debit and/or credit lines should be grouped and summed as follows in the export file:<ul><li>**Debit Accounts**: Company, Region, Super Fund, Cost Center Debit Number, Debit Account, Debit Account Sub, Fund Number, Project, Transaction Fee Account, and Location</li><li>**Credit Accounts**: Company, Region, Department, Super Fund, Cost Center Credit Number, Revenue Account, Revenue Account Sub Account, Fund Number, Project, and Location</li><li>**Financial Account Grouping**: Both credit and debit lines are grouped by Rock Financial Account, Project, and Transaction Fee Account.</li></ul> |
```
Comment (nateh777) [inline ShelbyFinancials/ReadMe.md:51]: "Revenue Account Sub Account" should just be "Revenue Account Sub"

File: ShelbyFinancials/ReadMe.md [hunk 8]
```diff
@@ -63,7 +63,7 @@ After install, the Shelby Financials Batch to Journal block was added to your Ba
...
-| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">4</span> | **GL Account Grouping** Determines if debit and/or credit lines should be grouped and summed by GL account in the export file |
+| <span style="width: 3em; height: 3em; line-height: 3em; background: #d21919; border-radius: 100%; color: white; text-align: center; display: inline-block;">3</span> | **GL Account Grouping** Determines if debit and/or credit lines should be grouped and summed as follows in the export file:<ul><li>**Debit Accounts**: Company, Region, Super Fund, Cost Center Debit Number, Debit Account, Debit Account Sub, Fund Number, Project, Transaction Fee Account, and Location</li><li>**Credit Accounts**: Company, Region, Department, Super Fund, Cost Center Credit Number, Revenue Account, Revenue Account Sub Account, Fund Number, Project, and Location</li><li>**Financial Account Grouping**: Both credit and debit lines are grouped by Rock Financial Account, Project, and Transaction Fee Account.</li></ul> |
```
Comment (nateh777) [inline ShelbyFinancials/ReadMe.md:66]: "Revenue Account Sub Account" should just be "Revenue Account Sub"

**Review by nateh777 — CHANGES_REQUESTED** [review]
Minor changes needed for documentation otherwise changes look good.

**Review by nateh777 — APPROVED** [review]

**Pull request merged**